hpl-2.2/man/man3/HPL_plindx10.3
.TH HPL_plindx10 3 "February 24, 2016" "HPL 2.2" "HPL Library Functions"
.SH NAME
HPL_plindx10 \- Compute the logarithmic maps for the spreading.
.SH SYNOPSIS
\fB\&#include "hpl.h"\fR
 
\fB\&void\fR
\fB\&HPL_plindx10(\fR
\fB\&HPL_T_panel *\fR
\fI\&PANEL\fR,
\fB\&const int\fR
\fI\&K\fR,
\fB\&const int *\fR
\fI\&IPID\fR,
\fB\&int *\fR
\fI\&IPLEN\fR,
\fB\&int *\fR
\fI\&IPMAP\fR,
\fB\&int *\fR
\fI\&IPMAPM1\fR
\fB\&);\fR
.SH DESCRIPTION
\fB\&HPL_plindx10\fR
computes  three arrays  IPLEN,  IPMAP  and  IPMAPM1  that
contain the logarithmic mapping information for the spreading phase.
.SH ARGUMENTS
.TP 8
PANEL   (local input/output)    HPL_T_panel *
On entry,  PANEL  points to the data structure containing the
panel information.
.TP 8
K       (global input)          const int
On entry, K specifies the number of entries in IPID.  K is at
least 2*N, and at most 4*N.
.TP 8
IPID    (global input)          const int *
On entry,  IPID  is an array of length K. The first K entries
of that array contain the src and final destination resulting
from the application of the interchanges.
.TP 8
IPLEN   (global output)         int *
On entry, IPLEN  is an array of dimension NPROW + 1. On exit,
this array is such that  IPLEN[i]  is the number of rows of A
in the processes  before process IMAP[i] after the sort, with
the convention that IPLEN[nprow] is the total number of rows.
In other words,  IPLEN[i+1] - IPLEN[i] is the local number of
rows of  A  that should be moved for each process.  IPLEN  is
such that the number of rows of the source process row can be
computed as IPLEN[1] - IPLEN[0], and the remaining entries of
this  array are sorted  so  that  the quantities IPLEN[i+1] -
IPLEN[i] are logarithmically sorted.
.TP 8
IPMAP   (global output)         int *
On entry, IPMAP is an array of dimension NPROW. On exit, this
array contains  the logarithmic mapping of the processes.  In
other words, IPMAP[myrow] is the corresponding sorted process
coordinate.
.TP 8
IPMAPM1 (global output)         int *
On entry, IPMAPM1  is an array of dimension NPROW.  On  exit,
this  array  contains  the inverse of the logarithmic mapping
contained  in  IPMAP:  IPMAPM1[ IPMAP[i] ] = i,  for all i in
[0.. NPROW)
.SH SEE ALSO
.BR HPL_pdlaswp00N \ (3),
.BR HPL_pdlaswp00T \ (3),
.BR HPL_pdlaswp01N \ (3),
.BR HPL_pdlaswp01T \ (3).


hpl-2.2/HISTORY
==============================================================
 High Performance Computing Linpack Benchmark (HPL)
 HPL - 2.2 - February 24, 2016
==============================================================

 History

 - 09/09/00 Public release of Version 1.0

 - 09/27/00 A couple of mistakes in the  VSIPL  port have been
 corrected.  The tar file as well as the web site were updated
 on September 27th, 2000.  Note  that  these problems were not
 affecting the BLAS version of the software in any way.

 - 01/01/04 Version 1.0a
 The  MPI  process grid  numbering  scheme  is now an run-time
 option.
 The inlined assembly  timer  routine that caused the compila-
 tion to fail when using  gcc  version 3.3  and above has been
 removed from the package.
 Various building problems on the T3E have been fixed;  Thanks
 to Edward Anderson.

 - 15/12/04 Version 1.0b
 Weakness of the pseudo-random matrix generator found for pro-
 blem sizes being power of twos and larger  than 2^15;  Thanks
 to Gregory Bauer. This problem has not been fixed. It is thus
 currently recommended to  HPL  users willing to test matrices
 of size larger than 2^15 to not use power twos.

 When the matrix size is such that one needs  > 16 GB  per MPI
 rank,  the  intermediate  calculation  (mat.ld+1) * mat.nq in
 HPL_pdtest.c  ends up  overflowing  because  it is done using
 32-bit arithmetic.  This issue has been fixed by  typecasting
 to size_t; Thanks to John Baron.

 - 09/10/08 Version 2.0

 Piotr Luszczek changed to 64-bit RNG, modified files:
 -- [M] include/hpl_matgen.h
 -- [M] testing/matgen/HPL_ladd.c
 -- [M] testing/matgen/HPL_lmul.c
 -- [M] testing/matgen/HPL_rand.c
 -- [M] testing/ptest/HPL_pdinfo.c

 For a motivation for the change, see:
    Dongarra and Langou, ``The Problem with the Linpack
    Benchmark Matrix Generator'', LAWN 206, June 2008.

 -- [M] testing/ptest/HPL_pdtest.c  --

 Julien Langou changed the test for correctness from
      ||Ax-b||_oo / ( eps * ||A||_1  * N            )
      ||Ax-b||_oo / ( eps * ||A||_1  * ||x||_1      )
      ||Ax-b||_oo / ( eps * ||A||_oo * ||x||_oo * N )
 to the normwise backward error
      || r ||_oo / ( eps * ( || x ||_oo * || A ||_oo + || b ||_oo ) * N )
 See:
  Nicholas J. Higham, ``Accuracy and Stability of Numerical Algorithms'',
  Society for Industrial and Applied Mathematics, Philadelphia, PA, USA,
  Second Edition, pages = xxx+680, ISBN = 0-89871-521-0, 2002.

 Note that in our case || b ||_oo is almost for sure
 1/2, we compute it anyway.

 - 10/26/2012 Version 2.1

 Piotr Luszczek introduced exact time stamping for HPL_pdgesv():
 -- [M] dist/include/hpl_misc.h
 -- [M] dist/testing/ptest/HPL_pdtest.c

 Piotr Luszczek fixed out-of-bounds access in data spreading functions
 and exact time stamping for HPL_pdgesv():
 -- [M] dist/src/pgesv/HPL_spreadN.c
 -- [M] dist/src/pgesv/HPL_spreadT.c
 Thanks to Stephen Whalen from Cray.

 - 02/24/2016 Version 2.2

 Piotr Luszczek added continuous reporting of factorization progress
 submitted by Intel and make scripts that uses Intel software tools and
 libraries and their Apple's Mac OS X equivalents.


hpl-2.2/src/pgesv/HPL_plindx10.c
/* 
 * -- High Performance Computing Linpack Benchmark (HPL)                
 *    HPL - 2.2 - February 24, 2016                          
 *    Antoine P. Petitet                                                
 *    University of Tennessee, Knoxville                                
 *    Innovative Computing Laboratory                                 
 *    (C) Copyright 2000-2008 All Rights Reserved                       
 *                                                                      
 * -- Copyright notice and Licensing terms:                             
 *                                                                      
 * Redistribution  and  use in  source and binary forms, with or without
 * modification, are  permitted provided  that the following  conditions
 * are met:                                                             
 *                                                                      
 * 1. Redistributions  of  source  code  must retain the above copyright
 * notice, this list of conditions and the following disclaimer.        
 *                                                                      
 * 2. Redistributions in binary form must reproduce  the above copyright
 * notice, this list of conditions,  and the following disclaimer in the
 * documentation and/or other materials provided with the distribution. 
 *                                                                      
 * 3. All  advertising  materials  mentioning  features  or  use of this
 * software must display the following acknowledgement:                 
 * This  product  includes  software  developed  at  the  University  of
 * Tennessee, Knoxville, Innovative Computing Laboratory.             
 *                                                                      
 * 4. The name of the  University,  the name of the  Laboratory,  or the
 * names  of  its  contributors  may  not  be used to endorse or promote
 * products  derived   from   this  software  without  specific  written
 * permission.                                                          
 *                                                                      
 * -- Disclaimer:                                                       
 *                                                                      
 * THIS  SOFTWARE  IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,  INCLUDING,  BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE UNIVERSITY
 * OR  CONTRIBUTORS  BE  LIABLE FOR ANY  DIRECT,  INDIRECT,  INCIDENTAL,
 * SPECIAL,  EXEMPLARY,  OR  CONSEQUENTIAL DAMAGES  (INCLUDING,  BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA OR PROFITS; OR BUSINESS INTERRUPTION)  HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT,  STRICT LIABILITY,  OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. 
 * ---------------------------------------------------------------------
 */ 
/*
 * Include files
 */
#include "hpl.h"

#ifdef STDC_HEADERS
void HPL_plindx10
(
   HPL_T_panel *                    PANEL,
   const int                        K,
   const int *                      IPID,
   int *                            IPLEN,
   int *                            IPMAP,
   int *                            IPMAPM1
)
#else
void HPL_plindx10
( PANEL, K, IPID, IPLEN, IPMAP, IPMAPM1 )
   HPL_T_panel *                    PANEL;
   const int                        K;
   const int *                      IPID;
   int *                            IPLEN;
   int *                            IPMAP;
   int *                            IPMAPM1;
#endif
{
/* 
 * Purpose
 * =======
 *
 * HPL_plindx10 computes  three arrays  IPLEN,  IPMAP  and  IPMAPM1  that
 * contain the logarithmic mapping information for the spreading phase.
 *
 * Arguments
 * =========
 *
 * PANEL   (local input/output)          HPL_T_panel *
 *         On entry,  PANEL  points to the data structure containing the
 *         panel information.
 *
 * K       (global input)                const int
 *         On entry, K specifies the number of entries in IPID.  K is at
 *         least 2*N, and at most 4*N.
 *
 * IPID    (global input)                const int *
 *         On entry,  IPID  is an array of length K. The first K entries
 *         of that array contain the src and final destination resulting
 *         from the application of the interchanges.
 *
 * IPLEN   (global output)               int *
 *         On entry, IPLEN  is an array of dimension NPROW + 1. On exit,
 *         this array is such that  IPLEN[i]  is the number of rows of A
 *         in the processes  before process IMAP[i] after the sort, with
 *         the convention that IPLEN[nprow] is the total number of rows.
 *         In other words,  IPLEN[i+1] - IPLEN[i] is the local number of
 *         rows of  A  that should be moved for each process.  IPLEN  is
 *         such that the number of rows of the source process row can be
 *         computed as IPLEN[1] - IPLEN[0], and the remaining entries of
 *         this  array are sorted  so  that  the quantities IPLEN[i+1] -
 *         IPLEN[i] are logarithmically sorted.
 *
 * IPMAP   (global output)               int *
 *         On entry, IPMAP is an array of dimension NPROW. On exit, this
 *         array contains  the logarithmic mapping of the processes.  In
 *         other words, IPMAP[myrow] is the corresponding sorted process
 *         coordinate.
 *
 * IPMAPM1 (global output)               int *
 *         On entry, IPMAPM1  is an array of dimension NPROW.  On  exit,
 *         this  array  contains  the inverse of the logarithmic mapping
 *         contained  in  IPMAP:  IPMAPM1[ IPMAP[i] ] = i,  for all i in
 *         [0.. NPROW)
 *
 * ---------------------------------------------------------------------
 */ 
/*
 * .. Local Variables ..
 */
   int                        dst, dstrow, i, ia, icurrow, jb, nb,
                              nprow, src, srcrow;
/* ..
 * .. Executable Statements ..
 */
   nprow = PANEL->grid->nprow; jb = PANEL->jb; nb = PANEL->nb;
   ia    = PANEL->ia;          icurrow = PANEL->prow;
/*
 * Compute  redundantly  the local number of rows  that each process has
 * and that belong to U in IPLEN[1 .. nprow+1]
 */
   for( i = 0; i <= nprow; i++ ) IPLEN[i] = 0;
 
   for( i = 0; i < K; i += 2 )
   {
      src = IPID[i]; Mindxg2p( src, nb, nb, srcrow, 0, nprow );
      if( srcrow == icurrow )
      {
         dst = IPID[i+1]; Mindxg2p( dst, nb, nb, dstrow, 0, nprow );
         if( ( dstrow != srcrow ) || ( dst - ia < jb ) ) IPLEN[dstrow+1]++;
      }
   }
/*
 * Logarithmic sort of the processes - compute IPMAP, IPLEN and IPMAPM1
 * (the inverse of IPMAP)
 */
   HPL_logsort( nprow, icurrow, IPLEN, IPMAP, IPMAPM1 );
/*
 * End of HPL_plindx10
 */
}
